### PR TITLE
Swap update

### LINKED ISSUE #2692: Issues with native swap
- [x] Reset amount when changing the token (reproduce: use send max on a token, then change the token to one with zero balance). via #2732
- [x] Prioritise source drop-down tokens same as WalletFragment; ie use Fiat value from high to low, then alphabetical order for the dropdown on source tokens. via #2724
- [x] Refresh swap price every 30 (?) seconds same as Uniswap does (check Uniswap maybe and time their refresh rate).
- [x] Add chain token to destination tokens, so we can swap to Eth (or Polygon etc).  via #2732
- [x] Expand swap provider message to include toolDetails.name provider from Quote. See //TODO: Expand swap provider. via #2732
- [x] Add the chain symbol after the 'fees', and also add fiat price similar to the gas widget. See //TODO: Needs to say 'Eth' via #2732

### PR DESCRIPTION
**Pull request by justindg**
Fixes #2692

- Update/fetch quotes every 30s
- Auto-retry fetch quote when encountering a timeout
- Display swap provider in quote summary + confirmation dialog
- Improve gas fee display in quote summary
- Clear source/dest amount values when switching source token
- Improve handling and display of error messages

### CODE REVIEW

**Comment by JamesSmartCell** [comment]
@justindg this is great, there's only one small issue I found:

While I have the confirmation dialog open, I don't want the 'Fetching Quote' update to interrupt it.

**Comment by seabornlee** [comment]
Awesome @justindg, some minor feedback, if you have any question writing unit test, please DM me.

**Comment by JamesSmartCell** [comment]
> Awesome @justindg, some minor feedback, if you have any questions about writing the unit test, please DM me.

Yes we will need to make writing unit tests for features a requirement for submitting PR. It may not be practical for some PRs but most allow for unit tests.

File: app/src/main/java/com/alphawallet/app/util/SwapUtils.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package com.alphawallet.app.util;
+
+import com.alphawallet.app.entity.lifi.Quote;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+
+public class SwapUtils
+{
+    public static final String GAS_PRICE_FORMAT = "%s %s";
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/util/SwapUtils.java:10]: These constants can be private.

File: app/src/main/java/com/alphawallet/app/util/SwapUtils.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    public static String getFormattedCurrentPrice(Quote quote)
+    {
+        return String.format(CURRENT_PRICE_FORMAT,
+                quote.action.fromToken.symbol,
+                getCurrentPrice(quote),
+                quote.action.toToken.symbol);
+    }
+
+    public static String getCurrentPrice(Quote quote)
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/util/SwapUtils.java:39]: This method can move to class `Quote` and convert to an instance method.

**Review by seabornlee — CHANGES_REQUESTED** [review]

**Review by JamesSmartCell — APPROVED** [review]

**Review by JamesSmartCell — APPROVED** [review]

**Comment by JamesSmartCell** [comment]
@seabornlee can you approve and merge?

**Review by seabornlee — APPROVED** [review]

**Comment by seabornlee** [comment]
Perfect! Thanks @justindg !

**Pull request merged**